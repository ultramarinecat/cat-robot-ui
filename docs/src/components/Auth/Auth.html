<!DOCTYPE html>

<html>
<head>
  <title>Auth.js</title>
  <meta http-equiv="content-type" content="text/html; charset=UTF-8">
  <meta name="viewport" content="width=device-width, target-densitydpi=160dpi, initial-scale=1.0; maximum-scale=1.0; user-scalable=0;">
  <link rel="stylesheet" media="all" href="../../../docco.css" />
</head>
<body>
  <div id="container">
    <div id="background"></div>
    
      <ul id="jump_to">
        <li>
          <a class="large" href="javascript:void(0);">Jump To &hellip;</a>
          <a class="small" href="javascript:void(0);">+</a>
          <div id="jump_wrapper">
          <div id="jump_page_wrapper">
            <div id="jump_page">
              
                
                <a class="source" href="../App/App.html">
                  src/components/App/App.js
                </a>
              
                
                <a class="source" href="../App/ErrorBoundary.html">
                  src/components/App/ErrorBoundary.js
                </a>
              
                
                <a class="source" href="../App/i18n.html">
                  src/components/App/i18n.js
                </a>
              
                
                <a class="source" href="Auth.html">
                  src/components/Auth/Auth.js
                </a>
              
                
                <a class="source" href="AuthActions.html">
                  src/components/Auth/AuthActions.js
                </a>
              
                
                <a class="source" href="AuthReducer.html">
                  src/components/Auth/AuthReducer.js
                </a>
              
                
                <a class="source" href="../Button/Button.html">
                  src/components/Button/Button.js
                </a>
              
                
                <a class="source" href="../Cat/Cat.html">
                  src/components/Cat/Cat.js
                </a>
              
                
                <a class="source" href="../Cat/CatActions.html">
                  src/components/Cat/CatActions.js
                </a>
              
                
                <a class="source" href="../Cat/CatReducer.html">
                  src/components/Cat/CatReducer.js
                </a>
              
                
                <a class="source" href="../Cat/fulltilt.html">
                  src/components/Cat/fulltilt.js
                </a>
              
                
                <a class="source" href="../Connection/Connection.html">
                  src/components/Connection/Connection.js
                </a>
              
                
                <a class="source" href="../Connection/ConnectionActions.html">
                  src/components/Connection/ConnectionActions.js
                </a>
              
                
                <a class="source" href="../Connection/ConnectionReducer.html">
                  src/components/Connection/ConnectionReducer.js
                </a>
              
                
                <a class="source" href="../Connection/messages.html">
                  src/components/Connection/messages.js
                </a>
              
                
                <a class="source" href="../Intro/Intro.html">
                  src/components/Intro/Intro.js
                </a>
              
                
                <a class="source" href="../Intro/IntroActions.html">
                  src/components/Intro/IntroActions.js
                </a>
              
                
                <a class="source" href="../Particles/Particles.html">
                  src/components/Particles/Particles.js
                </a>
              
                
                <a class="source" href="../Pulse/Pulse.html">
                  src/components/Pulse/Pulse.js
                </a>
              
                
                <a class="source" href="../Router/Router.html">
                  src/components/Router/Router.js
                </a>
              
                
                <a class="source" href="../Sound/CatSound.html">
                  src/components/Sound/CatSound.js
                </a>
              
                
                <a class="source" href="../Sound/Mute.html">
                  src/components/Sound/Mute.js
                </a>
              
                
                <a class="source" href="../Sound/Sound.html">
                  src/components/Sound/Sound.js
                </a>
              
                
                <a class="source" href="../Sound/TurningSound.html">
                  src/components/Sound/TurningSound.js
                </a>
              
                
                <a class="source" href="../Spinner/Spinner.html">
                  src/components/Spinner/Spinner.js
                </a>
              
                
                <a class="source" href="../Typing/Typing.html">
                  src/components/Typing/Typing.js
                </a>
              
                
                <a class="source" href="../../configureStore.html">
                  src/configureStore.js
                </a>
              
                
                <a class="source" href="../../index.html">
                  src/index.js
                </a>
              
                
                <a class="source" href="../../serviceWorker.html">
                  src/serviceWorker.js
                </a>
              
                
                <a class="source" href="../../setupTests.html">
                  src/setupTests.js
                </a>
              
            </div>
          </div>
        </li>
      </ul>
    
    <ul class="sections">
        
          <li id="title">
              <div class="annotation">
                  <h1>Auth.js</h1>
              </div>
          </li>
        
        
        
        <li id="section-1">
            <div class="annotation">
              
              <div class="pilwrap ">
                <a class="pilcrow" href="#section-1">&#182;</a>
              </div>
              <p>Auth component
Manages authenticating and connecting to robot</p>

            </div>
            
            <div class="content"><div class='highlight'><pre>
<span class="hljs-keyword">import</span> React, { memo, useState, useEffect, useCallback, useReducer, useRef } <span class="hljs-keyword">from</span> <span class="hljs-string">'react'</span>;
<span class="hljs-keyword">import</span> PropTypes <span class="hljs-keyword">from</span> <span class="hljs-string">'prop-types'</span>;
<span class="hljs-keyword">import</span> { connect } <span class="hljs-keyword">from</span> <span class="hljs-string">'react-redux'</span>;

<span class="hljs-keyword">import</span> { withTranslation } <span class="hljs-keyword">from</span> <span class="hljs-string">'react-i18next'</span>;
<span class="hljs-keyword">import</span> { isAlphanumeric, isByteLength } <span class="hljs-keyword">from</span> <span class="hljs-string">'validator'</span>;
<span class="hljs-keyword">import</span> classNames <span class="hljs-keyword">from</span> <span class="hljs-string">'classnames'</span>;

<span class="hljs-keyword">import</span> { connectToRobot } <span class="hljs-keyword">from</span> <span class="hljs-string">'./AuthActions.js'</span>;
<span class="hljs-keyword">import</span> Button <span class="hljs-keyword">from</span> <span class="hljs-string">'../Button/Button.js'</span>;
<span class="hljs-keyword">import</span> Spinner <span class="hljs-keyword">from</span> <span class="hljs-string">'../Spinner/Spinner.js'</span>;

<span class="hljs-keyword">import</span> styles <span class="hljs-keyword">from</span> <span class="hljs-string">'./Auth.module.scss'</span>;

<span class="hljs-keyword">const</span> CONNECTION_START = <span class="hljs-string">'CONNECTION_START'</span>;
<span class="hljs-keyword">const</span> CONNECTION_FAILED = <span class="hljs-string">'CONNECTION_FAILED'</span>;
<span class="hljs-keyword">const</span> INPUT_CHANGED = <span class="hljs-string">'INPUT_CHANGED'</span>;
<span class="hljs-keyword">const</span> INPUT_INVALID_ON_SUBMIT = <span class="hljs-string">'SUBMIT_INPUT_INVALID'</span>;

<span class="hljs-keyword">export</span> <span class="hljs-keyword">const</span> MIN_KEY_LENGTH = <span class="hljs-number">1</span>;
<span class="hljs-keyword">export</span> <span class="hljs-keyword">const</span> MAX_KEY_LENGTH = <span class="hljs-number">16</span>;

<span class="hljs-keyword">export</span> <span class="hljs-function"><span class="hljs-keyword">function</span> <span class="hljs-title">Auth</span>(<span class="hljs-params">{
  connecting,
  connectionTimedOut,
  connectionConflict,
  authFailed,
  connectToRobot,
  t
}</span>) </span>{
  <span class="hljs-keyword">const</span> startValidating = useRef(<span class="hljs-literal">false</span>);
  <span class="hljs-keyword">const</span> [catKey, setCatKey] = useState(<span class="hljs-string">''</span>);

  <span class="hljs-keyword">const</span> [
    { disableInput, disableButton, displaySpinner, feedback },
    dispatch
  ] = useReducer(authReducer, {
    <span class="hljs-attr">disableInput</span>: <span class="hljs-literal">false</span>,
    <span class="hljs-attr">disableButton</span>: <span class="hljs-literal">false</span>,
    <span class="hljs-attr">displaySpinner</span>: <span class="hljs-literal">false</span>,
    <span class="hljs-attr">feedback</span>: <span class="hljs-string">''</span>
  });

  <span class="hljs-keyword">const</span> handleInputChange = useCallback(<span class="hljs-function"><span class="hljs-params">event</span> =&gt;</span> {
    <span class="hljs-keyword">const</span> catKeyValue = event.target.value.trim();

    setCatKey(catKeyValue);</pre></div></div>
            
        </li>
        
        
        <li id="section-2">
            <div class="annotation">
              
              <div class="pilwrap ">
                <a class="pilcrow" href="#section-2">&#182;</a>
              </div>
              <p>don’t start validating until inital submit</p>

            </div>
            
            <div class="content"><div class='highlight'><pre>    <span class="hljs-keyword">if</span> (!startValidating.current) {
      <span class="hljs-keyword">return</span>;
    }

    dispatch({
      <span class="hljs-attr">type</span>: INPUT_CHANGED,
      <span class="hljs-attr">isValid</span>: isValidKey(catKeyValue),
      <span class="hljs-attr">feedback</span>: t(<span class="hljs-string">'auth_validation_cat_key'</span>)
    });
  });

  <span class="hljs-keyword">const</span> handleInputKeyPress = useCallback(<span class="hljs-function"><span class="hljs-params">event</span> =&gt;</span> {</pre></div></div>
            
        </li>
        
        
        <li id="section-3">
            <div class="annotation">
              
              <div class="pilwrap ">
                <a class="pilcrow" href="#section-3">&#182;</a>
              </div>
              <p>handle enter as submit click</p>

            </div>
            
            <div class="content"><div class='highlight'><pre>    <span class="hljs-keyword">if</span> (event.charCode === <span class="hljs-number">13</span>) {
      handleSubmit(event);
    }
  });

  <span class="hljs-keyword">const</span> handleSubmit = useCallback(<span class="hljs-function"><span class="hljs-params">event</span> =&gt;</span> {
    event.preventDefault();
    startValidating.current = <span class="hljs-literal">true</span>;

    <span class="hljs-keyword">if</span> (!isValidKey(catKey)) {
      dispatch({
        <span class="hljs-attr">type</span>: INPUT_INVALID_ON_SUBMIT,
        <span class="hljs-attr">feedback</span>: t(<span class="hljs-string">'auth_validation_cat_key'</span>)
      });

      <span class="hljs-keyword">return</span>;
    }</pre></div></div>
            
        </li>
        
        
        <li id="section-4">
            <div class="annotation">
              
              <div class="pilwrap ">
                <a class="pilcrow" href="#section-4">&#182;</a>
              </div>
              <p>if input valid, attempt to connect</p>

            </div>
            
            <div class="content"><div class='highlight'><pre>    dispatch({
      <span class="hljs-attr">type</span>: CONNECTION_START
    });

    connectToRobot(catKey);
  });

  useEffect(<span class="hljs-function"><span class="hljs-params">()</span> =&gt;</span> {
    <span class="hljs-keyword">if</span> (authFailed) {
      dispatch({
        <span class="hljs-attr">type</span>: CONNECTION_FAILED,
        <span class="hljs-attr">feedback</span>: t(<span class="hljs-string">'auth_invalid'</span>)
      });

      <span class="hljs-keyword">return</span>;
    }

    <span class="hljs-keyword">if</span> (connectionTimedOut) {
      dispatch({
        <span class="hljs-attr">type</span>: CONNECTION_FAILED,
        <span class="hljs-attr">feedback</span>: t(<span class="hljs-string">'auth_connection_failed'</span>)
      });

      <span class="hljs-keyword">return</span>;
    }

    <span class="hljs-keyword">if</span> (connectionConflict) {
      dispatch({
        <span class="hljs-attr">type</span>: CONNECTION_FAILED,
        <span class="hljs-attr">feedback</span>: t(<span class="hljs-string">'auth_connection_conflict'</span>)
      });
    }
  }, [connecting, connectionTimedOut, connectionConflict, authFailed]);

  <span class="hljs-keyword">const</span> labelClasses = classNames(styles.label, <span class="hljs-string">'col-sm-2'</span>, <span class="hljs-string">'d-none'</span>, <span class="hljs-string">'d-sm-block'</span>);
  <span class="hljs-keyword">const</span> feedbackClasses = classNames(
    styles.feedback,
    { <span class="hljs-attr">invisible</span>: !feedback },
    <span class="hljs-string">'col-sm-10'</span>,
    <span class="hljs-string">'offset-sm-2'</span>
  );

  <span class="hljs-keyword">return</span> (
    &lt;div className={styles.main}&gt;
      &lt;form className={styles.form} noValidate&gt;
        &lt;fieldset&gt;
          &lt;legend className={styles.legend}&gt;{t('auth_instructions')}&lt;/legend&gt;
          &lt;div className="row"&gt;
            &lt;label className={labelClasses} htmlFor="cat-key"&gt;
              {t('auth_cat_key')}
            &lt;/label&gt;
            &lt;div className="col-sm-10"&gt;
              &lt;input
                className="form-control"
                type="text"
                value={catKey}
                disabled={disableInput}
                onChange={handleInputChange}
                onKeyPress={handleInputKeyPress}
                id="cat-key"
                name="cat-key"
                placeholder={t('auth_placeholder')}
                required
              /&gt;
            &lt;/div&gt;
          &lt;/div&gt;
        &lt;/fieldset&gt;

        &lt;div className="row"&gt;
          &lt;div className={feedbackClasses} data-testid="feedback"&gt;
            {feedback}
          &lt;/div&gt;
        &lt;/div&gt;

        &lt;div&gt;
          &lt;Button
            text={t('auth_connect')}
            disabled={disableButton}
            onClick={handleSubmit}
            data-testid="button"
          /&gt;
          &lt;Spinner spinning={displaySpinner} data-testid="spinner" /&gt;
        &lt;/div&gt;
      &lt;/form&gt;
    &lt;/div&gt;
  );
}

export function authReducer(state, action) {
  switch (action.type) {
    case CONNECTION_START: {
      return {
        ...state,
        displaySpinner: true,
        disableButton: true,
        disableInput: true,
        feedback: ''
      };
    }
    case CONNECTION_FAILED: {
      const { feedback } = action;
      return {
        ...state,
        displaySpinner: false,
        disableButton: false,
        disableInput: false,
        feedback
      };
    }
    case INPUT_CHANGED:
      const { isValid, feedback } = action;
      return {
        ...state,
        disableButton: !isValid,
        feedback: !isValid ? feedback : ''
      };
    case INPUT_INVALID_ON_SUBMIT: {
      const { feedback } = action;
      return {
        ...state,
        disableButton: true,
        feedback
      };
    }
    default:
      return state;
  }
}

function isValidKey(key) {
  return (
    isByteLength(key, { min: MIN_KEY_LENGTH, max: MAX_KEY_LENGTH }) &amp;&amp; isAlphanumeric(key)
  );
}

Auth.propTypes = {
  connecting: PropTypes.bool,
  connectionTimedOut: PropTypes.bool,
  connectionConflict: PropTypes.bool,
  authFailed: PropTypes.bool,
  connectToRobot: PropTypes.func.isRequired,
  t: PropTypes.func.isRequired
};

const mapStateToProps = state =&gt; {
  return {
    connecting: state.auth.connecting,
    connectionTimedOut: state.auth.connectionTimedOut,
    connectionConflict: state.auth.connectionConflict,
    authFailed: state.auth.authFailed
  };
};

export default connect(
  mapStateToProps,
  { connectToRobot }
)(memo(withTranslation()(Auth)));</pre></div></div>
            
        </li>
        
    </ul>
  </div>
</body>
</html>
