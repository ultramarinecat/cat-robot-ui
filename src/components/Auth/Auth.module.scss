@charset 'utf-8';

////
//// Auth styles
////

@import 'base.scss';

// login form
.form {
  max-width: rem(560);
  @extend %center-block;
}

// login instructions
.legend {
  font-size: rem(18);
  margin-bottom: rem(16);
}

// input labels
.label.label {
  @extend %align-right;
  font-size: rem(14);
  padding-right: rem(10);
  padding-left: 0;
  margin-top: rem(7);
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
}

// form validation/login status
.feedback {
  @extend %align-left;
  font-size: rem(15);
  padding: rem(5) 0 rem(14) rem(18);
  min-height: rem(41);
  color: $validation-text-color;
}
