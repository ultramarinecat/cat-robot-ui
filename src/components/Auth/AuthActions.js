//
// Auth actions
//

export const CONNECT = 'CONNECT';

export function connectToRobot(catKey) {
  return {
    type: CONNECT,
    payload: catKey
  };
}
