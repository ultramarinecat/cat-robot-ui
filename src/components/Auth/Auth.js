//
// Auth component
// Manages authenticating and connecting to robot
//

import React, { memo, useState, useEffect, useCallback, useReducer, useRef } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import { withTranslation } from 'react-i18next';
import { isAlphanumeric, isByteLength } from 'validator';
import classNames from 'classnames';

import { connectToRobot } from './AuthActions.js';
import Button from '../Button/Button.js';
import Spinner from '../Spinner/Spinner.js';

import styles from './Auth.module.scss';

const CONNECTION_START = 'CONNECTION_START';
const CONNECTION_FAILED = 'CONNECTION_FAILED';
const INPUT_CHANGED = 'INPUT_CHANGED';
const INPUT_INVALID_ON_SUBMIT = 'SUBMIT_INPUT_INVALID';

export const MIN_KEY_LENGTH = 1;
export const MAX_KEY_LENGTH = 16;

export function Auth({
  connecting,
  connectionTimedOut,
  connectionConflict,
  authFailed,
  connectToRobot,
  t
}) {
  const startValidating = useRef(false);
  const [catKey, setCatKey] = useState('');

  const [
    { disableInput, disableButton, displaySpinner, feedback },
    dispatch
  ] = useReducer(authReducer, {
    disableInput: false,
    disableButton: false,
    displaySpinner: false,
    feedback: ''
  });

  const handleInputChange = useCallback(event => {
    const catKeyValue = event.target.value.trim();

    setCatKey(catKeyValue);

    // don't start validating until inital submit
    if (!startValidating.current) {
      return;
    }

    dispatch({
      type: INPUT_CHANGED,
      isValid: isValidKey(catKeyValue),
      feedback: t('auth_validation_cat_key')
    });
  });

  const handleInputKeyPress = useCallback(event => {
    // handle enter as submit click
    if (event.charCode === 13) {
      handleSubmit(event);
    }
  });

  const handleSubmit = useCallback(event => {
    event.preventDefault();
    startValidating.current = true;

    if (!isValidKey(catKey)) {
      dispatch({
        type: INPUT_INVALID_ON_SUBMIT,
        feedback: t('auth_validation_cat_key')
      });

      return;
    }

    // if input valid, attempt to connect
    dispatch({
      type: CONNECTION_START
    });

    connectToRobot(catKey);
  });

  useEffect(() => {
    if (authFailed) {
      dispatch({
        type: CONNECTION_FAILED,
        feedback: t('auth_invalid')
      });

      return;
    }

    if (connectionTimedOut) {
      dispatch({
        type: CONNECTION_FAILED,
        feedback: t('auth_connection_failed')
      });

      return;
    }

    if (connectionConflict) {
      dispatch({
        type: CONNECTION_FAILED,
        feedback: t('auth_connection_conflict')
      });
    }
  }, [connecting, connectionTimedOut, connectionConflict, authFailed]);

  const labelClasses = classNames(styles.label, 'col-sm-2', 'd-none', 'd-sm-block');
  const feedbackClasses = classNames(
    styles.feedback,
    { invisible: !feedback },
    'col-sm-10',
    'offset-sm-2'
  );

  return (
    <div className={styles.main}>
      <form className={styles.form} noValidate>
        <fieldset>
          <legend className={styles.legend}>{t('auth_instructions')}</legend>
          <div className="row">
            <label className={labelClasses} htmlFor="cat-key">
              {t('auth_cat_key')}
            </label>
            <div className="col-sm-10">
              <input
                className="form-control"
                type="text"
                value={catKey}
                disabled={disableInput}
                onChange={handleInputChange}
                onKeyPress={handleInputKeyPress}
                id="cat-key"
                name="cat-key"
                placeholder={t('auth_placeholder')}
                required
              />
            </div>
          </div>
        </fieldset>

        <div className="row">
          <div className={feedbackClasses} data-testid="feedback">
            {feedback}
          </div>
        </div>

        <div>
          <Button
            text={t('auth_connect')}
            disabled={disableButton}
            onClick={handleSubmit}
            data-testid="button"
          />
          <Spinner spinning={displaySpinner} data-testid="spinner" />
        </div>
      </form>
    </div>
  );
}

export function authReducer(state, action) {
  switch (action.type) {
    case CONNECTION_START: {
      return {
        ...state,
        displaySpinner: true,
        disableButton: true,
        disableInput: true,
        feedback: ''
      };
    }
    case CONNECTION_FAILED: {
      const { feedback } = action;
      return {
        ...state,
        displaySpinner: false,
        disableButton: false,
        disableInput: false,
        feedback
      };
    }
    case INPUT_CHANGED:
      const { isValid, feedback } = action;
      return {
        ...state,
        disableButton: !isValid,
        feedback: !isValid ? feedback : ''
      };
    case INPUT_INVALID_ON_SUBMIT: {
      const { feedback } = action;
      return {
        ...state,
        disableButton: true,
        feedback
      };
    }
    default:
      return state;
  }
}

function isValidKey(key) {
  return (
    isByteLength(key, { min: MIN_KEY_LENGTH, max: MAX_KEY_LENGTH }) && isAlphanumeric(key)
  );
}

Auth.propTypes = {
  connecting: PropTypes.bool,
  connectionTimedOut: PropTypes.bool,
  connectionConflict: PropTypes.bool,
  authFailed: PropTypes.bool,
  connectToRobot: PropTypes.func.isRequired,
  t: PropTypes.func.isRequired
};

const mapStateToProps = state => {
  return {
    connecting: state.auth.connecting,
    connectionTimedOut: state.auth.connectionTimedOut,
    connectionConflict: state.auth.connectionConflict,
    authFailed: state.auth.authFailed
  };
};

export default connect(
  mapStateToProps,
  { connectToRobot }
)(memo(withTranslation()(Auth)));
