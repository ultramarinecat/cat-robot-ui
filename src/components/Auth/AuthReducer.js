//
// Auth reducer
//

import {
  AUTH_SUCCESSFUL,
  AUTH_FAILED,
  AUTH_FORBIDDEN,
  CONNECTION_SUCCESSFUL,
  CONNECTION_TIMED_OUT,
  CONNECTION_CONFLICT
} from '../Connection/ConnectionActions.js';

import { CONNECT } from './AuthActions.js';

export default function authReducer(state = {}, action) {
  switch (action.type) {
    case CONNECT:
      return {
        ...state,
        ...{
          connecting: true,
          connectionTimedOut: false,
          connectionConflict: false,
          authFailed: false
        }
      };
    case AUTH_SUCCESSFUL:
      return {
        ...state,
        ...{
          authFailed: false,
          connecting: false
        }
      };
    case AUTH_FAILED:
      return {
        ...state,
        ...{
          authFailed: true,
          connecting: false
        }
      };
    case AUTH_FORBIDDEN:
      return {
        ...state,
        ...{
          authFailed: true,
          connecting: false
        }
      };
    case CONNECTION_SUCCESSFUL:
      return {
        ...state,
        ...{
          connectionTimedOut: false,
          connecting: false
        }
      };
    case CONNECTION_TIMED_OUT:
      return {
        ...state,
        ...{
          connectionTimedOut: true,
          connecting: false
        }
      };
    case CONNECTION_CONFLICT:
      return {
        ...state,
        ...{
          connectionConflict: true,
          connecting: false
        }
      };
    default:
      return state;
  }
}
