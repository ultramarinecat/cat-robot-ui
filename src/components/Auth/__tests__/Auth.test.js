import React from 'react';
import { render, fireEvent } from 'react-testing-library';

import { Auth, MIN_KEY_LENGTH, MAX_KEY_LENGTH } from '../Auth.js';

const CAT_KEY = 'auth_cat_key';
const CONNECT = 'auth_connect';

const INVALID_KEY = 'auth_validation_cat_key';
const INVALID_AUTH = 'auth_invalid';
const COULD_NOT_CONNECT = 'auth_connection_failed';
const CONNECTION_CONFLICT = 'auth_connection_conflict';

const SPINNER = 'spinner';
const FEEDBACK = 'feedback';

const HIDDEN = 'invisible';
const DISABLED = 'disabled';

describe('Auth', () => {
  it('should display a text field and a connect button', () => {
    const { container, getByLabelText, getByText } = render(
      <Auth connectToRobot={jest.fn()} t={key => key} />
    );

    expect(getByLabelText(CAT_KEY)).toBeInTheDocument();
    expect(getByText(CONNECT)).toBeInTheDocument();

    expect(container.firstChild).toMatchSnapshot();
  });

  it(`should initially not disable the input, disable the button, display a spinner,
      or display feedback`, () => {
    const { container, getByLabelText, getByText, getByTestId } = render(
      <Auth connectToRobot={jest.fn()} t={key => key} />
    );

    expect(getByLabelText(CAT_KEY)).not.toBeDisabled();
    expect(getByText(CONNECT)).not.toHaveClass(DISABLED);
    expect(getByTestId(SPINNER)).toHaveClass(HIDDEN);
    expect(getByTestId(FEEDBACK)).toHaveClass(HIDDEN);

    expect(container.firstChild).toMatchSnapshot();
  });

  it(`should disable the button and display invalid key feedback if the key is invalid
      when the button is clicked`, () => {
    const { container, getByLabelText, getByText, getByTestId } = render(
      <Auth connectToRobot={jest.fn()} t={key => key} />
    );

    const button = getByText(CONNECT);
    const input = getByLabelText(CAT_KEY);
    const feedback = getByTestId(FEEDBACK);

    fireEvent.change(input, { target: { value: 'cat_' } });
    fireEvent.click(button);

    expect(button).toHaveClass(DISABLED);
    expect(feedback).not.toHaveClass(HIDDEN);
    expect(feedback).toHaveTextContent(INVALID_KEY);

    expect(container.firstChild).toMatchSnapshot();
  });

  it(`should disable the button and display invalid key feedback whenever the key is invalid
      after the first button click`, () => {
    const { getByLabelText, getByText, getByTestId } = render(
      <Auth connectToRobot={jest.fn()} t={key => key} />
    );

    const button = getByText(CONNECT);
    const input = getByLabelText(CAT_KEY);
    const feedback = getByTestId(FEEDBACK);

    fireEvent.change(input, { target: { value: 'a_' } });

    expect(button).not.toHaveClass(DISABLED);
    expect(feedback).toHaveClass(HIDDEN);

    fireEvent.click(button);

    expect(button).toHaveClass(DISABLED);
    expect(feedback).not.toHaveClass(HIDDEN);
    expect(feedback).toHaveTextContent(INVALID_KEY);

    fireEvent.change(input, { target: { value: 'a'.repeat(MIN_KEY_LENGTH - 1) } });

    expect(button).toHaveClass(DISABLED);
    expect(feedback).not.toHaveClass(HIDDEN);
    expect(feedback).toHaveTextContent(INVALID_KEY);

    fireEvent.change(input, { target: { value: 'a'.repeat(MAX_KEY_LENGTH + 1) } });

    expect(button).toHaveClass(DISABLED);
    expect(feedback).not.toHaveClass(HIDDEN);
    expect(feedback).toHaveTextContent(INVALID_KEY);

    fireEvent.change(input, { target: { value: 'a'.repeat(MAX_KEY_LENGTH) } });

    expect(button).not.toHaveClass(DISABLED);
    expect(feedback).toHaveClass(HIDDEN);
  });

  it('should handle enter key press as a button click', () => {
    const { getByLabelText, getByText, getByTestId } = render(
      <Auth connectToRobot={jest.fn()} t={key => key} />
    );

    const button = getByText(CONNECT);
    const input = getByLabelText(CAT_KEY);
    const feedback = getByTestId(FEEDBACK);

    fireEvent.keyPress(input, { charCode: 13 });

    expect(button).toHaveClass(DISABLED);
    expect(feedback).not.toHaveClass(HIDDEN);
    expect(feedback).toHaveTextContent(INVALID_KEY);
  });

  it(`should display a spinner, disable the button, disable the input, hide feedback,
      and call the 'connectToRobot' prop with the cat key when the button is clicked`, () => {
    const catKey = 'cat';
    const connect = jest.fn();

    const { getByLabelText, getByText, getByTestId } = render(
      <Auth connectToRobot={connect} t={key => key} />
    );

    const button = getByText(CONNECT);
    const input = getByLabelText(CAT_KEY);
    const feedback = getByTestId(FEEDBACK);

    fireEvent.change(input, { target: { value: catKey } });
    fireEvent.click(button);

    expect(input).toBeDisabled(DISABLED);
    expect(button).toHaveClass(DISABLED);
    expect(feedback).toHaveClass(HIDDEN);
    expect(connect).toHaveBeenCalledWith(catKey);
  });

  it(`should hide the spinner, enable the button, enable the input, and display auth failed
      feedback when the auth fails`, () => {
    const { getByLabelText, getByText, getByTestId } = render(
      <Auth authFailed={true} connectToRobot={jest.fn()} t={key => key} />
    );

    const button = getByText(CONNECT);
    const input = getByLabelText(CAT_KEY);
    const feedback = getByTestId(FEEDBACK);
    const spinner = getByTestId(SPINNER);

    expect(spinner).toHaveClass(HIDDEN);
    expect(button).not.toHaveClass(DISABLED);
    expect(input).not.toBeDisabled();

    expect(feedback).not.toHaveClass(HIDDEN);
    expect(feedback).toHaveTextContent(INVALID_AUTH);
  });

  it(`should hide the spinner, enable the button, enable the input, and display connection
      timed out feedback when the auth fails when the connection times out`, () => {
    const { getByLabelText, getByText, getByTestId } = render(
      <Auth connectionTimedOut={true} connectToRobot={jest.fn()} t={key => key} />
    );

    const button = getByText(CONNECT);
    const input = getByLabelText(CAT_KEY);
    const feedback = getByTestId(FEEDBACK);
    const spinner = getByTestId(SPINNER);

    expect(spinner).toHaveClass(HIDDEN);
    expect(button).not.toHaveClass(DISABLED);
    expect(input).not.toBeDisabled();

    expect(feedback).not.toHaveClass(HIDDEN);
    expect(feedback).toHaveTextContent(COULD_NOT_CONNECT);
  });

  it(`should hide the spinner, enable the button, enable the input, and display connection
      conflict feedback when the auth fails when there is a connection conflict`, () => {
    const { getByLabelText, getByText, getByTestId } = render(
      <Auth connectionConflict={true} connectToRobot={jest.fn()} t={key => key} />
    );

    const button = getByText(CONNECT);
    const input = getByLabelText(CAT_KEY);
    const feedback = getByTestId(FEEDBACK);
    const spinner = getByTestId(SPINNER);

    expect(spinner).toHaveClass(HIDDEN);
    expect(button).not.toHaveClass(DISABLED);
    expect(input).not.toBeDisabled();

    expect(feedback).not.toHaveClass(HIDDEN);
    expect(feedback).toHaveTextContent(CONNECTION_CONFLICT);
  });
});
