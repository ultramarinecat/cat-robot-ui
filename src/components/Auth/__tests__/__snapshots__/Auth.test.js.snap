// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Auth should disable the button and display invalid key feedback if the key is invalid
      when the button is clicked 1`] = `
<div
  class="main waves-effect waves-light"
>
  <form
    class="form"
    novalidate=""
  >
    <fieldset>
      <legend
        class="legend"
      >
        auth_instructions
      </legend>
      <div
        class="row"
      >
        <label
          class="label col-sm-2 d-none d-sm-block"
          for="cat-key"
        >
          auth_cat_key
        </label>
        <div
          class="col-sm-10"
        >
          <input
            class="form-control"
            id="cat-key"
            name="cat-key"
            placeholder="auth_placeholder"
            required=""
            type="text"
            value="cat_"
          />
        </div>
      </div>
    </fieldset>
    <div
      class="row"
    >
      <div
        class="feedback col-sm-10 offset-sm-2"
        data-testid="feedback"
      >
        auth_validation_cat_key
      </div>
    </div>
    <div>
      <button
        class="main waves-effect waves-light disabled"
        type="button"
      >
        auth_connect
      </button>
      <div
        class="main invisible la-ball-clip-rotate la-sm waves-effect waves-light"
        data-testid="spinner"
      >
        <div />
      </div>
    </div>
  </form>
</div>
`;

exports[`Auth should display a text field and a connect button 1`] = `
<div
  class="main waves-effect waves-light"
>
  <form
    class="form"
    novalidate=""
  >
    <fieldset>
      <legend
        class="legend"
      >
        auth_instructions
      </legend>
      <div
        class="row"
      >
        <label
          class="label col-sm-2 d-none d-sm-block"
          for="cat-key"
        >
          auth_cat_key
        </label>
        <div
          class="col-sm-10"
        >
          <input
            class="form-control"
            id="cat-key"
            name="cat-key"
            placeholder="auth_placeholder"
            required=""
            type="text"
            value=""
          />
        </div>
      </div>
    </fieldset>
    <div
      class="row"
    >
      <div
        class="feedback invisible col-sm-10 offset-sm-2"
        data-testid="feedback"
      />
    </div>
    <div>
      <button
        class="main waves-effect waves-light"
        type="button"
      >
        auth_connect
      </button>
      <div
        class="main invisible la-ball-clip-rotate la-sm waves-effect waves-light"
        data-testid="spinner"
      >
        <div />
      </div>
    </div>
  </form>
</div>
`;

exports[`Auth should initially not disable the input, disable the button, display a spinner,
      or display feedback 1`] = `
<div
  class="main waves-effect waves-light"
>
  <form
    class="form"
    novalidate=""
  >
    <fieldset>
      <legend
        class="legend"
      >
        auth_instructions
      </legend>
      <div
        class="row"
      >
        <label
          class="label col-sm-2 d-none d-sm-block"
          for="cat-key"
        >
          auth_cat_key
        </label>
        <div
          class="col-sm-10"
        >
          <input
            class="form-control"
            id="cat-key"
            name="cat-key"
            placeholder="auth_placeholder"
            required=""
            type="text"
            value=""
          />
        </div>
      </div>
    </fieldset>
    <div
      class="row"
    >
      <div
        class="feedback invisible col-sm-10 offset-sm-2"
        data-testid="feedback"
      />
    </div>
    <div>
      <button
        class="main waves-effect waves-light"
        type="button"
      >
        auth_connect
      </button>
      <div
        class="main invisible la-ball-clip-rotate la-sm waves-effect waves-light"
        data-testid="spinner"
      >
        <div />
      </div>
    </div>
  </form>
</div>
`;
